### ISSUE TITLE
Allow building an Endless Key home directory

### ISSUE THREAD

**Issue by dbnicholson** [endlessm/eos-image-builder#117]
The image builder knows how to create a predefined Kolibri home at `/var/lib/kolibri` with a selection of content (described somewhere in our wondrous web of INI files). It generates a database file for Kolibri to pick up right where the image builder left off; no database migrations to worry about. We should change a lot of this. First, let's just get it working: where the image builder hard-codes `/var/lib/kolibri`, we want it to optionally use `/var/lib/endless-key`.

**Comment by jprvita** [comment]
I wrote a [draft PR](https://github.com/endlessm/eos-image-builder/pull/118) based on my initial understanding of the task description. But looking back at the project description and UX flow, I'm not sure how things are supposed to work: are we going to have a pre-defined list of pre-loaded content written-out in image configuration INI files, or have the user select the starter pack to be downloaded? The flow in the project spec has the starter pack selection screen, but isn't part of the idea to avoid the initial download for (some of) the Endless OS users?

**Comment by dbnicholson** [comment]
When I last spoke with @dylanmccall about this, I think the intention was to make the image builder use the content manifest files directly instead of converting to the image builder's INI format. However, I think we can start with the INI format if that's easier.

I could see adding a management command to kolibri-explore-plugin that just does the initial collection download immediately rather than the current method where it's driven by the frontend. Then the image builder configuration would be as simple as specifying the desired content pack. That would take a bit of work in the explore plugin, though.

**Comment by jprvita** [comment]
@dbnicholson Thanks, that helps clarify things a bit. However, on the current UX flow, for the English version of the app we're still expecting the user to select a starter pack, which to me seems to indicate that the image builder should not download / pre-load any content for the Endless Key app, since the user will choose it when they first run the app. That said, I don't know how well that model fits with the target audience of Endless OS, so I think we should check with @jofilizola.

With that said, you are right that we can start with pre-loading some content using the image builder configuration INI files, so I'll go ahead and create a test configuration and test my changes, and mark them ready for review once I'm happy with the code.

**Comment by dbnicholson** [comment]
Oh, I didn't realize the starter pack would be chosen at runtime. I guess one way forward would be preloading the content from all starter packs in the image. A slimmer alternative would be downloading all the channels and all the thumbnails from all the collections. You'd still need to download the actual content at runtime, but you could skip some of the other download steps.

**Comment by dbnicholson** [comment]
I think we can punt on the content right now. I think the most important things we needed to setup an Endless Key Kolibri homedir are:

* Enable/disable the proper plugins
* Download the apps bundle to a well known location
* Download the collections to a well known location
* Optionally preseed an `options.ini` file setting the apps bundle and collections paths rather than relying on environment variables to handle it

That's the bare minimum to have a system managed homedir supporting the explore plugin since it can already handle all the downloading at runtime.

**Comment by jprvita** [comment]
I had a quick call with @jofilizola earlier today and she clarified the intended behavior: the plan is to preload all Endless Key content into certain target images. There will still be a selection screen for a start pack (which maybe we should start calling something else), but that selection will only affect what appears in the discovery page. The library will show all the Endless Key content.

**Comment by jprvita** [comment]
> I think we can punt on the content right now. I think the most important things we needed to setup an Endless Key Kolibri homedir are:
> 
> * Enable/disable the proper plugins

I've got the list from https://github.com/endlessm/eos-administration/blob/master/ansible/roles/kolibri/templates/setup-kolibri.sh.j2#L109-L116 minus `pwa` and `kolibri_google_analytics_plugin`, which I think are only needed for the online instance -- please correct me if I'm wrong.

> * Download the apps bundle to a well known location
> * Download the collections to a well known location

So essentially the same as what is done in https://github.com/endlessm/eos-administration/blob/master/ansible/roles/kolibri/files/download-apps.sh.

> * Optionally preseed an `options.ini` file setting the apps bundle and collections paths rather than relying on environment variables to handle it

It looks like these paths already defaults to the right place if the options are not set. See https://github.com/endlessm/kolibri-explore-plugin/blob/master/kolibri_explore_plugin/views.py#L25-L28 and https://github.com/endlessm/kolibri-explore-plugin/blob/master/kolibri_explore_plugin/collectionviews.py#L33-L35

I have some draft code addressing all these points, pending testing.

**Comment by jprvita** [comment]
I have a [draft PR](https://github.com/endlessm/eos-image-builder/pull/119) ready for an initial round of review, however, I'm hitting the following problem when trying to enable `kolibri_explore_plugin` ([full logs](https://ci.endlessos.org/job/image-build-amd64/27425/console)):
```
++ 02:02:02 60-kolibri-content: kolibri plugin enable kolibri_explore_plugin
INFO     Enabling plugin 'kolibri_explore_plugin'
ERROR    Plugin 'kolibri_explore_plugin' exists but does not have an importable kolibri_plugin module
Error: One or more plugins could not be enabled
```

**Comment by dbnicholson** [comment]
> I have a [draft PR](https://github.com/endlessm/eos-image-builder/pull/119) ready for an initial round of review, however, I'm hitting the following problem when trying to enable `kolibri_explore_plugin` ([full logs](https://ci.endlessos.org/job/image-build-amd64/27425/console)):
> 
> ```
> ++ 02:02:02 60-kolibri-content: kolibri plugin enable kolibri_explore_plugin
> INFO     Enabling plugin 'kolibri_explore_plugin'
> ERROR    Plugin 'kolibri_explore_plugin' exists but does not have an importable kolibri_plugin module
> Error: One or more plugins could not be enabled
> ```

I think this is because Kolibri 0.15.7 is being used:

```
++ 02:01:42 60-kolibri-content: pip install kolibri==0.15.7
Collecting kolibri==0.15.7
  Using cached kolibri-0.15.7-py2.py3-none-any.whl (100.8 MB)
Installing collected packages: kolibri
Successfully installed kolibri-0.15.7
```

The current explore plugin releases need the 0.16 and the error is because the plugin API changed. You'll have to go the app route and install from a github wheel like https://github.com/learningequality/kolibri/releases/download/v0.16.0-beta5/kolibri-0.16.0b5-py2.py3-none-any.whl.

This might be too messed up to carry on right now. The idea here is that you want to use a Kolibri version less than or equal to the one that will be in the Flatpak. Otherwise you risk running database migrations that code in the flatpak doesn't know about. However, neither the Kolibri flatpak nor the Endless Key flatpak are actually using Kolibri 0.16 yet. See https://github.com/endlessm/endless-key-flatpak/pull/9.

What would be a little more appropriate, particularly since `60-kolibri-content` isn't actually doing any caching anymore, is to do all of this using the flatpak itself. However, I don't know if that's feasible in the image builder. You can see how eos-kolibri runs `kolibri` [in the flatpak](https://github.com/endlessm/eos-kolibri/blob/3529eb964d598ca21baf1b74150a5ff1851b4146/src/eos-kolibri-manage.in#L37) if you want to give it a try. This would all have to be a chroot hook, though.

**Comment by jprvita** [comment]
I got back to this for a bit and wrote the changes to use kolibri from the flatpak, but I'm now hitting an error from `libglnx` ([full logs](https://ci.endlessos.org/view/Image/job/image-build-amd64/27491/console)):
```
+ '[' org.learningequality.Kolibri == org.learningequality.Kolibri ']'
+ PLUGINS_DISABLE=()
+ PLUGINS_ENABLE=("kolibri.plugins.app" "kolibri_app_desktop_xdg_plugin" "kolibri_desktop_auth_plugin")
+ export KOLIBRI_HOME=/var/lib/kolibri
+ KOLIBRI_HOME=/var/lib/kolibri
+ INSTALL_CHANNELS=("${KOLIBRI_INSTALL_CHANNELS[@]}")
+ KOLIBRI_CMD='flatpak run --no-desktop     --filesystem=/var/lib/kolibri --command=/app/bin/kolibri org.learningequality.Kolibri'
+ for plugin in "${PLUGINS_ENABLE[@]}"
+ flatpak run --no-desktop --filesystem=/var/lib/kolibri --command=/app/bin/kolibri org.learningequality.Kolibri plugin enable kolibri.plugins.app
**
flatpak:ERROR:subprojects/libglnx/glnx-shutil.c:155:mkdir_p_at_internal: assertion failed: (!did_recurse)
Bail out! flatpak:ERROR:subprojects/libglnx/glnx-shutil.c:155:mkdir_p_at_internal: assertion failed: (!did_recurse)
/tmp/60-kolibri-content.chroot: line 98: 24294 Aborted                 ${KOLIBRI_CMD} plugin enable ${plugin}
```
(I'm running `flatpak run` as root here to avoid having to deal with creating the `kolibri` user for now, but I don't think it is affecting the problem I'm seeing here).

**Comment by jprvita** [comment]
> (I'm running `flatpak run` as root here to avoid having to deal with creating the `kolibri` user for now, but I don't think it is affecting the problem I'm seeing here).

The `kolibri` user in created EOS through a [systemd sysusers snippet](https://github.com/endlessm/eos-kolibri/blob/master/data/sysusers.d/eos-kolibri.conf.in), and its home directory permissions are adjusted through a [systemd tmpfiles snippet](https://github.com/endlessm/eos-kolibri/blob/master/data/tmpfiles.d/eos-kolibri.conf.in), so we don't need to worry about doing that in the image builder -- we can create KOLIBRI_HOME and pre-download channels in the image builder as `root`, and leave the permissions to be adjusted on first boot (we will need to extend that tmpfiles snippet to cover the Endless Key home too).

**Comment by jprvita** [comment]
> I got back to this for a bit and wrote the changes to use kolibri from the flatpak, but I'm now hitting an error from `libglnx` ([full logs](https://ci.endlessos.org/view/Image/job/image-build-amd64/27491/console)):
> 
> ```
> + '[' org.learningequality.Kolibri == org.learningequality.Kolibri ']'
> + PLUGINS_DISABLE=()
> + PLUGINS_ENABLE=("kolibri.plugins.app" "kolibri_app_desktop_xdg_plugin" "kolibri_desktop_auth_plugin")
> + export KOLIBRI_HOME=/var/lib/kolibri
> + KOLIBRI_HOME=/var/lib/kolibri
> + INSTALL_CHANNELS=("${KOLIBRI_INSTALL_CHANNELS[@]}")
> + KOLIBRI_CMD='flatpak run --no-desktop     --filesystem=/var/lib/kolibri --command=/app/bin/kolibri org.learningequality.Kolibri'
> + for plugin in "${PLUGINS_ENABLE[@]}"
> + flatpak run --no-desktop --filesystem=/var/lib/kolibri --command=/app/bin/kolibri org.learningequality.Kolibri plugin enable kolibri.plugins.app
> **
> flatpak:ERROR:subprojects/libglnx/glnx-shutil.c:155:mkdir_p_at_internal: assertion failed: (!did_recurse)
> Bail out! flatpak:ERROR:subprojects/libglnx/glnx-shutil.c:155:mkdir_p_at_internal: assertion failed: (!did_recurse)
> /tmp/60-kolibri-content.chroot: line 98: 24294 Aborted                 ${KOLIBRI_CMD} plugin enable ${plugin}
> ```

I'm still hitting this error when trying to call `kolibri` from inside the flatpak. And as we discussed in today's meeting, we don't strictly need to enable the plugins, so I can avoid the previous error I was hitting when trying to enable kolibri-explore-plugin (https://github.com/endlessm/eos-image-builder/issues/117#issuecomment-1707562704) by simply skipping it for now. I believe that is fine and should get us a working KOLIBRI_HOME, so I'll stick with the pip-installed kolibri.

**Comment by jprvita** [comment]
I also looked into adjusting the permissions of KOLIBRI_HOME for Endless Key, and found out that it is already addressed by https://github.com/endlessm/eos-kolibri/commit/d9d441944430b0fcb23c78eb1b63a339bb624b6f. However, we do not create a separate system user for Endless Key, so its KOLIBRI_HOME will be chown'ed to `kolibri:kolibri`. I believe that is fine.

**Comment by jprvita** [comment]
This is ready for review with https://github.com/endlessm/eos-image-builder/pull/119. A test image that installs all our Spanish content can be found at https://images.endlessos.org/files/nightly/eoscustom-amd64-amd64/master/ektest/230915-211737/.

@dbnicholson @dylanmccall Now, https://github.com/endlessm/endless-key-flatpak/pull/62 makes me think that perhaps we should not be doing automatic provisioning in the image builder either? Let me know if I should drop that from the PR.

**Comment by dbnicholson** [comment]
> @dbnicholson @dylanmccall Now, [endlessm/endless-key-flatpak#62](https://github.com/endlessm/endless-key-flatpak/pull/62) makes me think that perhaps we should not be doing automatic provisioning in the image builder either? Let me know if I should drop that from the PR.

@dylanmccall would know better, but I think if you can get away without provisioning the device you shouldn't do it. I was thinking that it needs to be provisioned in order to import content, but I actually don't think that's the case. In other words, if we can just import content and leave everything else until runtime, that's better. It's been a while since I was manually importing content on a fresh Kolibri homedir.

**Comment by dylanmccall** [comment]
> > @dbnicholson @dylanmccall Now, [endlessm/endless-key-flatpak#62](https://github.com/endlessm/endless-key-flatpak/pull/62) makes me think that perhaps we should not be doing automatic provisioning in the image builder either? Let me know if I should drop that from the PR.
> 
> @dylanmccall would know better, but I think if you can get away without provisioning the device you shouldn't do it. I was thinking that it needs to be provisioned in order to import content, but I actually don't think that's the case. In other words, if we can just import content and leave everything else until runtime, that's better. It's been a while since I was manually importing content on a fresh Kolibri homedir.

Yes, that's correct: don't do it :b

I think you've got it all already, but just to be sure, the more detailed answer is there's two pieces here :) I think you're right that we can import content without having a facility set up. But we _do_ need Kolibri in the image builder to build a database, with all the migrations run and that big glob of channel metadata and imported content so the final system knows what it's working with. (We should be able to do away with that in the future by making cleverer use of content manifest files, but that isn't happening yet). In the image builder we end up running `kolibri manage deprovision` to be extra sure that the default Kolibri database we have created is _effectively_ starting from scratch, except for the content marked as available.

The other piece of this is we can get the image builder to add a file to Kolibri's home directory, named `automatic_provision.json`. As you noted from https://github.com/endlessm/endless-key-flatpak/pull/62, this indeed doesn't do anything in the Endless Key app, and I don't think it'll do anything in the next version of the Kolibri app. We used that file for one particular custom image, but only because they had some additional requirements. At this point, the settings we used for that image are actually the settings we use _everywhere_ now.

**Comment by jprvita** [comment]
> > > @dbnicholson @dylanmccall Now, [endlessm/endless-key-flatpak#62](https://github.com/endlessm/endless-key-flatpak/pull/62) makes me think that perhaps we should not be doing automatic provisioning in the image builder either? Let me know if I should drop that from the PR.
> > 
> > 
> > @dylanmccall would know better, but I think if you can get away without provisioning the device you shouldn't do it. I was thinking that it needs to be provisioned in order to import content, but I actually don't think that's the case. In other words, if we can just import content and leave everything else until runtime, that's better. It's been a while since I was manually importing content on a fresh Kolibri homedir.
> 
> Yes, that's correct: don't do it :b
> 
> I think you've got it all already, but just to be sure, the more detailed answer is there's two pieces here :) I think you're right that we can import content without having a facility set up. But we _do_ need Kolibri in the image builder to build a database, with all the migrations run and that big glob of channel metadata and imported content so the final system knows what it's working with. (We should be able to do away with that in the future by making cleverer use of content manifest files, but that isn't happening yet). In the image builder we end up running `kolibri manage deprovision` to be extra sure that the default Kolibri database we have created is _effectively_ starting from scratch, except for the content marked as available.
> 
> The other piece of this is we can get the image builder to add a file to Kolibri's home directory, named `automatic_provision.json`. As you noted from [endlessm/endless-key-flatpak#62](https://github.com/endlessm/endless-key-flatpak/pull/62), this indeed doesn't do anything in the Endless Key app, and I don't think it'll do anything in the next version of the Kolibri app. We used that file for one particular custom image, but only because they had some additional requirements. At this point, the settings we used for that image are actually the settings we use _everywhere_ now.

What I was referring to in my previous point was the `automatic_provision.json` file, which seems we don't need to generate for the Endless Key, so I have dropped that part from my PR. I'll leave it as-is for Kolibri for now, but it sounds like we will want to drop it for Kolibri too at some point. I don't think we are doing any other kind of provisioning. We are still enabling the plugins, importing the channels, and running `kolibri manage deprovision` at the end of it all, for each separate KOLIBRI_HOME (Kolibri and Endless Key). The only other thing is creating a `options.ini` file setting `REGULAR_USERS_CAN_MANAGE_CONTENT`, which I'm doing for both KOLIBRI_HOME dirs as well.

**Comment by jprvita** [comment]
Now, I've built an image with both Kolibri and Endless Key installed side-by-side, and while the right channels seem to be available in the right places (opening the Kolibri app only shows the channels I've installed for Kolibri, and there are launchers for the other channels on the last page of the icon grid), things are not working properly: the Endless Key app opens but shows a "Server Error (500)", and any of the desktop launchers, including the ones for channels installed in Kolibri's KOLIBRI_HOME (ex., Cooking) are launched within Endless Key.

I don't know if these problems are related to the pre-populated home dir or are always happening when trying to run Kolibri and Endless Key side-by-side: I haven't had a chance to test it outside of a preloaded image yet.

![Screenshot from 2023-09-18 14-34-20](https://github.com/endlessm/eos-image-builder/assets/73894/55dd83e4-d8f4-4af5-aca4-7342498ef92d)

![Screenshot from 2023-09-18 14-33-37](https://github.com/endlessm/eos-image-builder/assets/73894/de180161-fcd8-4636-9655-a4a3ea813958)
![Screenshot from 2023-09-18 14-37-59](https://github.com/endlessm/eos-image-builder/assets/73894/91a4faba-39bc-47d5-9650-ddae318668b4)

**Comment by dbnicholson** [comment]
The 500 seems bad, likely a crash in the backend. I would definitely look in the journal for that one.

I noticed the other day that both Endless Key and Kolibri use the same custom `x-kolibri-dispatch://` scheme from kolibri-app-desktop-xdg-plugin for launching channels. So, if you have both apps installed, the channel could be opened in either app. We may want to make Endless Key the default in eos-theme so it takes precedence if both are installed on EOS. Or use a different scheme for Endless Key so it's unambiguous.

**Comment by dylanmccall** [comment]
> The 500 seems bad, likely a crash in the backend. I would definitely look in the journal for that one.
> 
> I noticed the other day that both Endless Key and Kolibri use the same custom `x-kolibri-dispatch://` scheme from kolibri-app-desktop-xdg-plugin for launching channels. So, if you have both apps installed, the channel could be opened in either app. We may want to make Endless Key the default in eos-theme so it takes precedence if both are installed on EOS. Or use a different scheme for Endless Key so it's unambiguous.

Yeah, we should rename that URI scheme. It's also used for channel shortcuts in the launcher. There's also `kolibri:` and `x-kolibri-app:`. Those ones shouldn't cause an issue (and there's some related code which can probably be trimmed upstream), but there should be no harm renaming them if they are.

**Comment by jprvita** [comment]
> The 500 seems bad, likely a crash in the backend. I would definitely look in the journal for that one.

You are right. Looks like `list(cls.registered_hooks)` is empty when the code did not expect it to be.

[endless-key-daemon-crash.txt](https://github.com/endlessm/eos-image-builder/files/12667065/endless-key-daemon-crash.txt)

**Comment by dbnicholson** [comment]
> > The 500 seems bad, likely a crash in the backend. I would definitely look in the journal for that one.
> 
> You are right. Looks like `list(cls.registered_hooks)` is empty when the code did not expect it to be.
> 
> [endless-key-daemon-crash.txt](https://github.com/endlessm/eos-image-builder/files/12667065/endless-key-daemon-crash.txt)

I think this is because in the image builder you've disabled `kolibri.plugins.default_theme` and enabled `kolibri_endless_key_theme`, but we're not currently including or enabling `kolibri_endless_key_theme` in the flatpak. So, at runtime there's no theme at all and it falls over.

We should probably do that in endless-key-flatpak. For now I'd just drop that part. To be honest, I don't think touching the plugins is really needed at all since the flatpak can handle it. All we really need is core kolibri so that we can preseed content and let everything else happen at runtime.

**Comment by jprvita** [comment]
> > > The 500 seems bad, likely a crash in the backend. I would definitely look in the journal for that one.
> > 
> > 
> > You are right. Looks like `list(cls.registered_hooks)` is empty when the code did not expect it to be.
> > [endless-key-daemon-crash.txt](https://github.com/endlessm/eos-image-builder/files/12667065/endless-key-daemon-crash.txt)
> 
> I think this is because in the image builder you've disabled `kolibri.plugins.default_theme` and enabled `kolibri_endless_key_theme`, but we're not currently including or enabling `kolibri_endless_key_theme` in the flatpak. So, at runtime there's no theme at all and it falls over.
> 
> We should probably do that in endless-key-flatpak. For now I'd just drop that part. To be honest, I don't think touching the plugins is really needed at all since the flatpak can handle it. All we really need is core kolibri so that we can preseed content and let everything else happen at runtime.

Ok, I'll work on reducing what is done in the image builder and see if that resolves this problem.

Now on HOW to do that...

My initial understanding from the latest feedback on https://github.com/endlessm/eos-image-builder/pull/119/files is that it would be preferable to do something like `kolibri importcontent --manifest spanish.json network` and magically all the channel ids and node ids from spanish.json would be imported. But reading through kolibri's code, it seems that importcontent still expects a channel id (but maybe I'm misreading it). And the discussion about putting the content somewhere else on disk before importing made me question my initial understanding even further (although without making it any clearer).

@dbnicholson @dylanmccall Can you help clarify what approach should we try next? I'm still not super familiar with working with Kolibri content, and the documentation seems to indicate that I would still need to iterate over a list of channels, which does not seem the same as "using our JSON collection files" mean. Thanks!

**Comment by dbnicholson** [comment]
Sorry about all of this. What I thought was needed isn't actually what's needed. When this started I thought we'd need to get the shared home directory pretty close to provisioned. But with some recent modifications, the Endless Key flatpak (and probably soon the Kolibri flatpak) can do almost all of the setup itself.

So, this really becomes a task about importing content. And then at the end doing `deprovision` just to be safe. As to importing the content, I think there's really just one way right now:

* For each pack in the collections JSON embedded in the flatpak, read out the desired channels and node IDs and import them with `importchannel` and `importcontent`.

In a shinier future, there could hopefully be an `importmanifest` type of command. That doesn't exist today. Furthermore, since the cache server needs to be instructed to fetch all the nodes for some channel, you have to dig the channels out of the collections JSON files.

In python it would be pretty straightforward to parse the JSON. You could probably do it with `jq`, too:

```
$ jq -r '.channels[].id' /var/lib/flatpak/app/org.endlessos.Key/current/active/files/share/endless-key/collections/artist-0001.json 
1520f018610256549c98ca0140cceebe
59b8deeb90f544da923187e77c8d3820
057f871caa405ec29d62ba0523c193d7
f061fce103ff5d4e9b8433e67802e666
e11462f71c6f5472b113311c69071b05
b40491d1ef8b5506b8c6ae861372e9de
79a50be66bad5eb686c42617c914fd45
1b1fc9bd453a4c52bb5628d9ae804ede
97111903de564de49483a9705d41a8ac
f1ada7abc4194ff48a958337a31972c7
$ jq -r '.channels[] | select(.id == "59b8deeb90f544da923187e77c8d3820") | .include_node_ids[]' /var/lib/flatpak/app/org.endlessos.Key/current/active/files/share/endless-key/collections/artist-0001.json 
18794db672544db9b1fc5113b0f11ab0
8b094975fb034ffabe8f84e94236c450
454b964a232b40b6b1ac757bc8682442
bedd4f11617a4a91bea8eb563eade367
b6a27f3179e545a7ae36aba9ebcdd7e0
```

**Comment by dbnicholson** [comment]
One thing I remembered about the preseeding is that (I believe) they want all the content from every channel in every collection. So, I don't think you have to bother getting the `include_node_ids` out. Just get the channel IDs and import them in full.

In shell, I'd probably do it something like this:

```
export KOLIBRI_HOME="${OSTREE_VAR}/lib/endless-key/data"
if [ -n "${EIB_KOLIBRI_CENTRAL_CONTENT_BASE_URL}" ]; then
  export KOLIBRI_CENTRAL_CONTENT_BASE_URL="${EIB_KOLIBRI_CENTRAL_CONTENT_BASE_URL}"
fi
collections=${OSTREE_VAR}/lib/flatpak/app/org.endlessos.Key/current/active/files/share/endless-key/collections/*.json
for collection in $collections; do
  jq -r '.channels[].id'  "$collection" >> /tmp/channels
done
channels=$(sort -u /tmp/channels)
rm -f /tmp/channels
for channel in $channels; do
  kolibri manage importchannel network "$channel"
  kolibri manage importcontent --fail-on-error network "$channel"
done
kolibri manage deprovision
```

The remaining parts are:

* Figure out an appropriate kolibri version to install locally. Maybe we just hardcode a version known to be less than what's in the flatpak like we're doing now.
* Tell the cache server to import those channels ahead of time. Since you won't know the channels until the flatpak is installed (and you can therefore read the collections JSON), you may need to copy/move `hooks/content/50-kolibri-content` into the `image` stage after flatpaks are installed.

Since the process doesn't uses collections instead of image builder config for the content configuration, I kinda feel like you should make new hooks. I don't think we should worry too much about duplication at the moment.

**Comment by jprvita** [comment]
I believe what I have written now is doing the right thing. In addition to what is described on the last comments above, I'm also allowing the user to specify in the configuration which collections (artist, explorer etc) they want to preload, although I'm not entirely sure if we'll use it or not -- depends on how we end up defining the preloading strategy. I can easily drop that part if we don't want it.

But I am hitting a problem when preloading the content, where `importchannel` is failing:
```
++ 02:19:44 53-ek-content-preload: kolibri manage importchannel network 97111903de564de49483a9705d41a8ac
INFO     Option DEBUG in section [Server] being overridden by environment variable KOLIBRI_DEBUG
INFO     Option DEBUG_LOG_DATABASE in section [Server] being overridden by environment variable KOLIBRI_DEBUG_LOG_DATABASE
INFO     Option CENTRAL_CONTENT_BASE_URL in section [Urls] being overridden by environment variable KOLIBRI_CENTRAL_CONTENT_BASE_URL
WARNING  Neither `ifconfig` (`ifconfig -a`) nor `ip` (`ip address show`) commands are available, listing network interfaces is likely to fail
INFO     Running Kolibri with the following settings: kolibri.deployment.default.settings.base
ERROR    /bin/sh: 1: ip: not found

INFO     Importing 'tasks' module from django apps
INFO     Invoking command importchannel network 97111903de564de49483a9705d41a8ac
INFO     Downloading data for channel id 97111903de564de49483a9705d41a8ac
Traceback (most recent call last):
  File "/var/cache/eos-image-builder/tmp/kolibri-content-venv/lib/python3.9/site-packages/kolibri/dist/sqlalchemy/engine/base.py", line 1283, in _execute_context
    self.dialect.do_execute(
  File "/var/cache/eos-image-builder/tmp/kolibri-content-venv/lib/python3.9/site-packages/kolibri/dist/sqlalchemy/engine/default.py", line 590, in do_execute
    cursor.execute(statement, parameters)
sqlite3.DatabaseError: database disk image is malformed
```
In my tests it always fails on that same channel, `97111903de564de49483a9705d41a8ac`.

I've opened a [draft PR](https://github.com/endlessm/eos-image-builder/pull/120) with my changes.

**Comment by jprvita** [comment]
> In my tests it always fails on that same channel, `97111903de564de49483a9705d41a8ac`.

Trying with collections that do not include that channel (athlete + spanish) [got me to avoid that problem](https://ci.endlessos.org/view/Image/job/image-build-amd64/27647/consoleFull), although the build failed for an unrelated reason (good old `partx -a -v /dev/loop0 failed 10 times; giving up`).

**Comment by dbnicholson** [comment]
> In my tests it always fails on that same channel, `97111903de564de49483a9705d41a8ac`.

I imagine the channel database is corrupted on the content server. Let me take a look.

**Comment by dbnicholson** [comment]
I deleted the channel on the server, so the next time you run this it should be imported again.

**Comment by jprvita** [comment]
> I deleted the channel on the server, so the next time you run this it should be imported again.

Great, thanks! I'll give it a try. That may have happened because I aborted one build while it was running the hook that instructs the server to download some channels, but it is just a guess. Let's see if it builds successfully now.

In any case, I was able to test a build with the athlete + spanish collections, and it looks like things are working.

![Screenshot from 2023-09-27 13-31-43](https://github.com/endlessm/eos-image-builder/assets/73894/7884024b-be89-4af3-8af5-8c74ee54d002)

There are a few things we may want to address separately though.

1. The discovery page is empty.  
![Screenshot from 2023-09-27 13-30-50](https://github.com/endlessm/eos-image-builder/assets/73894/218b3bc7-75ed-401b-9204-d429db9d6c99)

2. After running the Endless Key for the first time, desktop lauchers appear for its channels.
![Screenshot from 2023-09-27 13-32-59](https://github.com/endlessm/eos-image-builder/assets/73894/cede0c99-1f50-4984-accf-1b54e062271e)

3. Clicking on one of these launchers works, and it gets opened in the Endless Key, except that the window title and icon are set to the channel name and icon. And after clicking on the `x` inside the application (next to the channel icon), there is a prompt to choose which app to use.
![Screenshot from 2023-09-27 13-41-20](https://github.com/endlessm/eos-image-builder/assets/73894/bfc1edf9-856e-4f30-bfa5-dc1db4646661)
![Screenshot from 2023-09-27 13-41-44](https://github.com/endlessm/eos-image-builder/assets/73894/4fad436c-8478-47eb-818d-2ff17597132b)

4. Clicking on one of the Kolibri launchers also opens them in the Endless Key app, which works or not depending on the channel (see below "Cooking" and "Using Studio: Your Content Workspace for Kolibri").
![Screenshot from 2023-09-27 13-44-46](https://github.com/endlessm/eos-image-builder/assets/73894/acde77bb-3f04-49e9-9307-dc5be91f7b0f)
![Screenshot from 2023-09-27 14-17-06](https://github.com/endlessm/eos-image-builder/assets/73894/6a2562db-08c1-4c38-a40e-98ad9510102f)

**Comment by jprvita** [comment]
> > I deleted the channel on the server, so the next time you run this it should be imported again.
> 
> Great, thanks! I'll give it a try. That may have happened because I aborted one build while it was running the hook that instructs the server to download some channels, but it is just a guess. Let's see if it builds successfully now.

[Same error](https://ci.endlessos.org/view/Image/job/image-build-amd64/27649/console), with the same channel ID:
```
[37mINFO     Invoking command importchannel network 97111903de564de49483a9705d41a8ac[0m
[37mINFO     Downloading data for channel id 97111903de564de49483a9705d41a8ac[0m
Traceback (most recent call last):
  File "/var/cache/eos-image-builder/tmp/kolibri-content-venv/lib/python3.9/site-packages/kolibri/dist/sqlalchemy/engine/base.py", line 1283, in _execute_context
    self.dialect.do_execute(
  File "/var/cache/eos-image-builder/tmp/kolibri-content-venv/lib/python3.9/site-packages/kolibri/dist/sqlalchemy/engine/default.py", line 590, in do_execute
    cursor.execute(statement, parameters)
sqlite3.DatabaseError: database disk image is malformed
```

**Comment by jprvita** [comment]
> > > I deleted the channel on the server, so the next time you run this it should be imported again.
> > 
> > 
> > Great, thanks! I'll give it a try. That may have happened because I aborted one build while it was running the hook that instructs the server to download some channels, but it is just a guess. Let's see if it builds successfully now.
> 
> [Same error](https://ci.endlessos.org/view/Image/job/image-build-amd64/27649/console), with the same channel ID:
> 
> ```
> �[37mINFO     Invoking command importchannel network 97111903de564de49483a9705d41a8ac�[0m
> �[37mINFO     Downloading data for channel id 97111903de564de49483a9705d41a8ac�[0m
> Traceback (most recent call last):
>   File "/var/cache/eos-image-builder/tmp/kolibri-content-venv/lib/python3.9/site-packages/kolibri/dist/sqlalchemy/engine/base.py", line 1283, in _execute_context
>     self.dialect.do_execute(
>   File "/var/cache/eos-image-builder/tmp/kolibri-content-venv/lib/python3.9/site-packages/kolibri/dist/sqlalchemy/engine/default.py", line 590, in do_execute
>     cursor.execute(statement, parameters)
> sqlite3.DatabaseError: database disk image is malformed
> ```

And it looks like the channel was re-imported (other channels just jumped from 0 straight to 100%):
```
+ 20:44:13 INFO 52-ek-content-cache: Importing channel 97111903de564de49483a9705d41a8ac content
+ 20:44:13 INFO 52-ek-content-cache: Progress: 0%
+ 20:44:17 INFO 52-ek-content-cache: Progress: 5%
+ 20:44:20 INFO 52-ek-content-cache: Progress: 11%
+ 20:44:22 INFO 52-ek-content-cache: Progress: 17%
+ 20:44:23 INFO 52-ek-content-cache: Progress: 21%
+ 20:44:23 INFO 52-ek-content-cache: Progress: 25%
+ 20:44:25 INFO 52-ek-content-cache: Progress: 30%
+ 20:44:26 INFO 52-ek-content-cache: Progress: 35%
+ 20:44:28 INFO 52-ek-content-cache: Progress: 41%
+ 20:44:29 INFO 52-ek-content-cache: Progress: 45%
+ 20:44:30 INFO 52-ek-content-cache: Progress: 50%
+ 20:44:32 INFO 52-ek-content-cache: Progress: 55%
+ 20:44:34 INFO 52-ek-content-cache: Progress: 60%
+ 20:44:35 INFO 52-ek-content-cache: Progress: 65%
+ 20:44:37 INFO 52-ek-content-cache: Progress: 70%
+ 20:44:39 INFO 52-ek-content-cache: Progress: 77%
+ 20:44:40 INFO 52-ek-content-cache: Progress: 80%
+ 20:44:43 INFO 52-ek-content-cache: Progress: 85%
+ 20:44:45 INFO 52-ek-content-cache: Progress: 90%
+ 20:44:46 INFO 52-ek-content-cache: Progress: 97%
+ 20:44:48 INFO 52-ek-content-cache: Progress: 100%
```

**Comment by jprvita** [comment]
I believe the import problem is unrelated to the changes. These hooks are executed unless two things are true: A. `org.endlessos.Key` is part of the `apps` list under `[flatpak-remote-flathub]` in the image configuration; and B. the `collections` list under `[endlesskey]` in the image configuration is not empty; so I changed the default so `collections` is now empty by default. With that, I think this is ready for review + merging, so I'm marking the [PR](https://github.com/endlessm/eos-image-builder/pull/120) as ready for review.

**Comment by dbnicholson** [comment]
I built this locally with the `artist` collection (which contains channel `97111903de564de49483a9705d41a8ac`) and wasn't able to reproduce the problem. I also locally imported the channel with both kolibri 0.15.12 and 0.16.0-beta6 and didn't have an issue with either. So, I don't know what's going on there.

One potential issue is that we're doing the installation with 0.15.7, which is very old even for the nearly dead 0.15 series. I'm going to add on a couple commits that:

* Update the kolibri version to 0.15.12 since org.learningequality.Kolibri has had that for a while.
* Specifies a separate kolibri version for Endless Key since the Kolibri version there is not connected to the version used in the Kolibri flatpak. I'm going to use 0.16.0-beta6 since that's been in the released Key flatpak.

**Comment by dbnicholson** [comment]
> 1. The discovery page is empty.

We still need to select a content pack so that the tags can be applied at runtime.

https://github.com/endlessm/kolibri-explore-plugin/issues/863

> 2. After running the Endless Key for the first time, desktop lauchers appear for its channels.
> 3. Clicking on one of these launchers works, and it gets opened in the Endless Key, except that the window title and icon are set to the channel name and icon. And after clicking on the `x` inside the application (next to the channel icon), there is a prompt to choose which app to use.
> 4. Clicking on one of the Kolibri launchers also opens them in the Endless Key app, which works or not depending on the channel (see below "Cooking" and "Using Studio: Your Content Workspace for Kolibri").

The product decision was to disable the generated channel launchers for Endless Key, so these will go away.

https://github.com/endlessm/eos-kolibri/issues/29

**Comment by dbnicholson** [comment]
I merged #120. @jprvita can you backport this to the eos5.0 branch? Then I guess we need to setup some configuration that actually tries to use it. But I believe we may be blocked on the proper collection to use.

**Comment by jprvita** [comment]
I've created #121 addressing the deduplication of the list of channels to be downloaded. Once that is merged I can test with eos5.0 and backport.

**Comment by jprvita** [comment]
> > 4. Clicking on one of the Kolibri launchers also opens them in the Endless Key app, which works or not depending on the channel (see below "Cooking" and "Using Studio: Your Content Workspace for Kolibri").
> 
> The product decision was to disable the generated channel launchers for Endless Key, so these will go away.
> 
> [endlessm/eos-kolibri#29](https://github.com/endlessm/eos-kolibri/issues/29)

Will those changes also prevent number 4 above? Or do we need to remove a mime type handler somewhere else?

**Comment by dylanmccall** [comment]
> > > 4. Clicking on one of the Kolibri launchers also opens them in the Endless Key app, which works or not depending on the channel (see below "Cooking" and "Using Studio: Your Content Workspace for Kolibri").
> > 
> > 
> > The product decision was to disable the generated channel launchers for Endless Key, so these will go away.
> > [endlessm/eos-kolibri#29](https://github.com/endlessm/eos-kolibri/issues/29)
> 
> Will those changes also prevent number 4 above? Or do we need to remove a mime type handler somewhere else?

Oh, for some reason I thought we'd filed that one away somewhere, but I can't find it. I made a separate ticket for it over here: https://github.com/endlessm/endless-key-flatpak/issues/102. It's a minor fix we need to do in the Endless Key desktop app, independent of ostree stuff.

**Comment by dbnicholson** [comment]
This work needs to be backported to the eos5.0 branch, although it would be nice to give it a test on master first.

**Comment by jprvita** [comment]
I have backported this to eos5.0 and tested a 5.0 build. Things seem to be working, but it still tried to "download" content when I selected the Explorer pack (which was already preloaded). However, it was very fast (probably about 1 minute), and looking at the logs it seems only the tagging was happening at that time. It would be nice to fix the UX so it does not look like to the user that there is still some download happening.

Here is the config I used for testing:
```
[flatpak-remote-flathub]
apps_add =
  org.endlessos.Key

[kolibri]
automatic_provision = true
regular_users_can_manage_content = true
install_channels =
  # Cooking [complete]
  79cd09863eed51e98576c35ede6f9c9d
  # Using Studio: Your Content Workspace for Kolibri [6]
  ec29f4cc20a8437d844a60297c2ffd07

[kolibri-ec29f4cc20a8437d844a60297c2ffd07]
include_node_ids =
  # English [topic]
  8b0c58bf187a4b88ba9bd602302b5736

[endlesskey]
collections =
  artist
  explorer
```

And here are the logs: 
[endless-key-import-logs.txt](https://github.com/endlessm/eos-image-builder/files/12888094/endless-key-import-logs.txt)

Since the changes here look correct, I've pushed them to eos5.0.

**Comment by erikos** [comment]
This is followed-up by https://github.com/endlessm/kolibri-explore-plugin/issues/890
